packages/transformers/js/src/visitors/modules.js

```diff
@@ -0,0 +1,376 @@
+        if (declaration.id != null && isIdentifier(declaration.id)) {
+          let id = t.identifier(declaration.id.name);
+          state.scope.addReference(id);
+
+          return [
+            declaration,
+            t.expressionStatement(
+              t.assignmentExpression(
+                '=',
+                t.memberExpression(
+                  t.identifier('exports'),
+                  t.identifier('default'),
+                ),
+                id,
+              ),
+            ),
+          ];
+        } else if (declaration.type !== 'TSDeclareFunction') {
+          // Replace with an assignment to `exports`.
+          return t.expressionStatement(
+            t.assignmentExpression(
+              '=',
+              t.memberExpression(
+                t.identifier('exports'),
+                t.identifier('default'),
+              ),
+              t.toExpression(declaration),
+            ),
+          );
+        } else {
+          return REMOVE;
+        }
+      };
+    },
+  },
+  ThisExpression(node, {scope}) {
+    while (scope) {
+      if (scope.type === 'function') {
+        return;
+      }
+
+      scope = scope.parent;
+    }
+
+    return t.identifier('undefined');
+  },
+};
+
+function getNames(state, source) {
+  let names = state.importNames.get(source.value);
+  if (!names) {
+    let name = state.scope.generateUid(source.value);
+    names = {name, namespace: '', default: ''};
+    state.importNames.set(source.value, names);
+  }
+
+  return names;
+}
+
+function getSpecifier(state, source, name) {
+  let names = getNames(state, source);
+  return t.memberExpression(t.identifier(names.name), t.identifier(name));
+}
+
+function getDefault(state, source) {
+  let names = getNames(state, source);
+  if (!names.default) {
+    names.default = state.scope.generateUid(names.name + 'Default');
+  }
+  return t.identifier(names.default);
+}
+
+function getNamespace(state, source) {
+  let names = getNames(state, source);
+  if (!names.namespace) {
+    names.namespace = state.scope.generateUid(names.name + 'Namespace');
+  }
+  return t.identifier(names.namespace);
+}
+
+const visitor = mergeVisitors(scopeVisitor, modulesVisitor);
+
+export function esm2cjs(ast: BabelNodeFile, asset?: MutableAsset) {
+  let imports = [];
+  let importNames = new Map();
+  let scope = new Scope('program');
+  let exports = [];
+  let state: State = {
+    imports,
+    importNames,
+    exports,
+    needsInteropFlag: false,
+    scope,
+  };
+
+  traverse2(ast, visitor, state);
+
+  let body = ast.program.body;
+  let prepend = [];
+  let helpersId;
+  let addHelpers = () => {
+    if (helpersId) {
+      return helpersId;
+    }
+
+    // Add a dependency so Parcel includes the helpers
+    let helperPath = path.join(__dirname, '..', 'esmodule-helpers.js');
+    let helperSpecifier = asset
+      ? relativePath(path.dirname(asset.filePath), helperPath)
```

Comment:
Could this instead be `@parcel/transformer-js/something/esmodule-helpers.js`? This relative path might contain a machine-specific part

Reply:
This is probably causing this Windows test failure:
```
1) babel
     should support building with default babel config when running parcel globally:
   Error: Unknown pipeline ./D.
    at ResolverRunner.resolve (D:\a\1\s\packages\core\core\src\requests\/PathRequest.js:143:17)
    at Object.run (D:\a\1\s\packages\core\core\src\requests\/PathRequest.js:66:20)
    at RequestTracker.runRequest (D:\a\1\s\packages\core\core\src\/RequestTracker.js:535:20)
    at AssetGraphBuilder.runPathRequest (D:\a\1\s\packages\core\core\src\/AssetGraphBuilder.js:765:18)
    at PromiseQueue._runFn (D:\a\1\s\packages\core\utils\src\/PromiseQueue.js:84:7)
    at PromiseQueue._next (D:\a\1\s\packages\core\utils\src\/PromiseQueue.js:73:5)
```

Reply:
hmm tried this, and it mostly works, except in a few tests that are libraries where "includeNodeModules" is implicitly false. Then the helpers aren't actually included in the bundle... I don't think we can get away with that because `@parcel/transformer-js` won't be a dependency of the package. Any ideas?

Reply:
I didn't think of that... and Yarn PnP would disallow it as well

Reply:
So go back to the relative path you think?

Reply:
Yeah, we don't have an alternative at the moment...

Reply:
Ok, actually ended up going back to the non-relative path because of one test that runs parcel "globally" which happens to be on a different drive letter on windows. To make this work, I had to add a `resolveFrom` option to dependencies so it would resolve relative to the plugin and not the source file, and also override the `env` of the dependency to always include node modules for that file. Currently `resolveFrom` is in addition to `sourceFile` (which wasn't publicly writable), and it falls back to `sourceFile` when not available. I'm not sure if these fields should be combined...